Marlin/src/lcd/ultralcd.cpp

```diff
@@ -902,20 +902,21 @@ void MarlinUI::update() {
         // When reversing the encoder direction, a movement step can be missed because
         // encoderDiff has a non-zero residual value, making the controller unresponsive.
         // The fix clears the residual value when the encoder is reversed.
-        // Also check if passed half the threshold. This will compensate for missed single steps.
+        // Also check if past half the threshold to compensate for missed single steps.
         static int8_t lastEncoderDiff;
+        int8_t prevDiff = lastEncoderDiff;
+        lastEncoderDiff = encoderDiff;  // Store before updating encoderDiff to save actual steps
+
         // When not past threshold, and reversing... or past half the threshold
-        if ((encoderDiff != 0 && abs_diff < (ENCODER_PULSES_PER_STEP)) && // Not past threshold
-            (
-              abs_diff > (ENCODER_PULSES_PER_STEP) / 2 || // Passed half the threshold
-              (ABS(encoderDiff - lastEncoderDiff) >= (ENCODER_PULSES_PER_STEP) && ABS(lastEncoderDiff) < (ENCODER_PULSES_PER_STEP)) // Or direction change through zero
-            )) {
-            lastEncoderDiff = encoderDiff; // Store before updating encoderDiff, we want actual steps
-            encoderDiff = (encoderDiff < 0 ? -1 : 1) * (ENCODER_PULSES_PER_STEP); // Treat as full step
-            abs_diff = ENCODER_PULSES_PER_STEP;
-        }
-        else {
-          lastEncoderDiff = encoderDiff;
+        if (WITHIN(abs_diff, 1, (ENCODER_PULSES_PER_STEP) - 1)  // Not past threshold
+          && (abs_diff > (ENCODER_PULSES_PER_STEP) / 2          // Passed half the threshold?
+            || (ABS(encoderDiff - prevDiff) >= (ENCODER_PULSES_PER_STEP)  // Or a large change...
```

Comment:
This line is not a test for a 'large change', but rather to detect a direction change. This is the actual bugfix! It replaces the '<' and '>' tests, which were not correct in all cases.

Reply:
This individual test is non-specific, even though in context it means "reverse." In fact, it would also catch the case of a sudden large change without a reverse. But, the next condition is what makes it more specific.

Reply:
Right, but this condition can only become true when stepping through zero, and thus changing direction.
This is because in all other situations (ABS(encoderDiff - prevDiff) >= (ENCODER_PULSES_PER_STEP) && ABS(prevDiff) < (ENCODER_PULSES_PER_STEP)) cannot become true here. This is not obvious.

Reply:
I'm fine to label it as "definitely a reverse" but because it is non-obvious I want to also include a comment to point out why it _must_ be a reverse, and why it can't be a sudden large difference due to a very fast spin of the encoder.